pkg/xds/experimental/client/client.go

```diff
@@ -0,0 +1,399 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package xdsclient
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/cilium/cilium/pkg/backoff"
+	"github.com/cilium/cilium/pkg/envoy/xds"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+
+	corepb "github.com/cilium/proxy/go/envoy/config/core/v3"
+	discoverypb "github.com/cilium/proxy/go/envoy/service/discovery/v3"
+)
+
+// requestCons is a constraint for request messages in xDS protocol.
+type requestCons interface {
+	*discoverypb.DiscoveryRequest | *discoverypb.DeltaDiscoveryRequest
+}
+
+// responseCons is a constraint for response messages in xDS protocol.
+type responseCons interface {
+	*discoverypb.DiscoveryResponse | *discoverypb.DeltaDiscoveryResponse
+	GetTypeUrl() string
+}
+
+// transport is a common generic interface of xDS gRPC client.
+type transport[req requestCons, resp responseCons] interface {
+	Send(req) error
+	Recv() (resp, error)
+}
+
+// nameToResource maps a resource name to a proto representation of the resource.
+type nameToResource map[string]proto.Message
+
+// tx represents a transaction prepared based on {Delta,}DiscoveryResponse.
+type tx struct {
+	typeUrl string
+	updated nameToResource
+	deleted []string
+}
+
+type txs []tx
+
+// getter specifies function to retrieve all resources of given type url.
+type getter func(typeUrl string) (*xds.VersionedResources, error)
+
+// flavour specifies a common interface for implementations specific to sotw or
+// delta protocol version. It is primarily used by XDSClient.
+type flavour[ReqT requestCons, RespT responseCons] interface {
```

Comment:
Are we sure we need generics here vs just plain interfaces? I remember discussing this with Aleksander but don't remember the conclusion. The use of generics makes it so all of the code needs to use generics.